### DISCUSSION TITLE
How to implement Multiple-Instances-Pattern for UserTask

### DISCUSSION THREAD

**Discussion by experdot** [elsa-workflows/elsa-core#1539] [Q&A]
An activity is able to initiate multiple instances of itself when triggered. See more:
[Pattern 13 (Multiple Instances with a priori Design-Time Knowledge)](http://www.workflowpatterns.com/patterns/control/multiple_instance/wcp13.php)

Can elsa-core do that?

**Comment by sfmskywalker** [comment]
Elsa can't do this yet, but it will be supported as of Elsa 3. By the way, that's a great resource you linked to that I wasn't aware of before. Thanks for your question!

**Reply by Nokecy** [reply]
Wait for this ~~~

**Comment by experdot** [comment]
I tried the SimpleForkWorkfow sample for MultipleInstancePattern but failed.

```
    public class SimpleForkWorkflow : IWorkflow
    {
        public void Build(IWorkflowBuilder builder)
        {
            builder
                .WriteLine("This demonstrates a simple workflow with switch.")
                .WriteLine("Using switch we can branch a workflow.")
                .Then<ForkBranchDecisionActivity>(fork =>
                {
                    var users = new string[] { "john", "tom", "jenny" };
                    var joinMode = Join.JoinMode.WaitAll;
                    foreach (var user in users)
                    {
                        fork.When("a").WriteLine("An fork created for " + user)
                            .Then<Join>(join => join.WithMode(joinMode))
                            .ThenNamed("Finish");
                    }
                })
                .WriteLine("Workflow finished.").WithName("Finish");
        }
    }
```

Expected output:
```
Typing 'a' will result in going through branch A
Any other key will result in branch B
a
An fork created for jenny
An fork created for tom
An fork created for john
Workflow finished.
```

Actual output:
```
Typing 'a' will result in going through branch A
Any other key will result in branch B
a
An fork created for jenny
Workflow finished.
An fork created for tom
Workflow finished.
An fork created for john
Workflow finished.
```

What should I do to achieve the expected result?

**Reply by sfmskywalker** [reply]
The following workflow will provide you with the expected output:

```csharp
/// <summary>
/// Demonstrates connecting outcomes of activities to other activities.
/// </summary>
public class SimpleForkWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        var users = new[] { "john", "tom", "jenny" };

        builder
            .WriteLine("This demonstrates a simple workflow with switch.")
            .WriteLine("Using switch we can branch a workflow.")
            .Then<Fork>(fork => fork.WithBranches(users), fork =>
            {
                foreach (var user in users)
                {
                    fork.When(user)
                        .WriteLine("A fork created for " + user)
                        .ThenNamed("Join1");
                }
            })
            .Then<Join>(join => join.WithMode(Join.JoinMode.WaitAll)).WithName("Join1")
            .WriteLine("Workflow finished.");
    }
}
```

**Reply by Nokecy** [reply]
If `users` are acquired dynamically, where should they be processed @sfmskywalker

**Reply by Nokecy** [reply]
`
public class SimpleForkWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        var users = new[] { "john", "tom", "jenny" };

        builder
            .WriteLine("This demonstrates a simple workflow with switch.")
            .WriteLine("Using switch we can branch a workflow.")
            .Then<Fork>(fork => fork.WithBranches(users), fork =>
            {

                    //How to get Branches Property values

                foreach (var user in users)
                {
                    fork.When(user)
                        .WriteLine("A fork created for " + user)
                        .ThenNamed("Join1");
                }
            })
            .Then<Join>(join => join.WithMode(Join.JoinMode.WaitAll)).WithName("Join1")
            .WriteLine("Workflow finished.");
    }
}
`

**Reply by sfmskywalker** [reply]
Elsa 2 doesn't let you change the workflow definition at runtime, because that would be similar to changing C# statements at runtime.

Probably a more suitable activity to use would be the `ForEach` activity, allowing you to iterate over a dynamic list of values (e.g. coming from a database).

Example:

```csharp
public class SimpleForkWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        var users = new[] { "john", "tom", "jenny" };

        builder
            .WriteLine("This demonstrates a simple workflow with switch.")
            .WriteLine("Using switch we can branch a workflow.")
            .ForEach(users, body =>
            {
                body
                    .WriteLine(context =>
                    {
                        // Get current item.
                        var user = context.ForEachScope<string>();
                        return "Handling user" + user;
                    });
            })
            .WriteLine("Workflow finished.");
    }
}
```

**Reply by Nokecy** [reply]
Maybe I should wait #1059 @sfmskywalker
